File: 0241/solution.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> list[int]:
        def apply_op(a, b, op):
            if op == "+":
                return a + b
            elif op == "-":
                return a - b
            elif op == "*":
                return a * b

        def parse_expression(expression):
            nums, ops = [], []
            num = 0
            for i, c in enumerate(expression):
                if c.isdigit():
                    num = num * 10 + int(c)
                else:
                    nums.append(num)
                    ops.append(c)
                    num = 0
            nums.append(num)  # Append the last number
            return nums, ops

        # Recursively compute the possible results for the expression
        def solve(left, right):
            if (left, right) in memo:
                return memo[(left, right)]

            # Base case: if there is only one number, return it
            if left == right:
                return [nums[left]]

            results = []
            # Iterate over the operators in the current segment
            for i in range(left, right):
                # Recursively compute the results of the left and right subexpressions
                left_results = solve(left, i)
                right_results = solve(i + 1, right)

                # Combine the results of left and right subexpressions using the current operator
                for left_res in left_results:
                    for right_res in right_results:
                        results.append(apply_op(left_res, right_res, ops[i]))

            # Memoize the computed results for the current segment
            memo[(left, right)] = results
            return results

        nums, ops = parse_expression(expression)  # Parse the expression
        memo = {}  # Dictionary for memoization
        return solve(0, len(nums) - 1)  # Solve for the entire expression

File: 0241/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private List<Integer> numbers;
    private List<Character> operators;
    private List<Integer>[][] dp;

    public List<Integer> diffWaysToCompute(String expression) {
        parseExpression(expression);
        int n = numbers.size();
        dp = new ArrayList[n][n];
        return compute(0, n - 1);
    }

    private List<Integer> compute(int left, int right) {
        if (dp[left][right] != null) {
            return dp[left][right];
        }

        List<Integer> result = new ArrayList<>();
        if (left == right) {
            result.add(numbers.get(left));
        } else {
            for (int i = left; i < right; i++) {
                List<Integer> leftPart = compute(left, i);
                List<Integer> rightPart = compute(i + 1, right);
                char op = operators.get(i);

                for (int l : leftPart) {
                    for (int r : rightPart) {
                        result.add(calculate(l, r, op));
                    }
                }
            }
        }

        dp[left][right] = result;
        return result;
    }

    private int calculate(int a, int b, char op) {
        if (op == '+') {
            return a + b;
        }
        if (op == '-') {
            return a - b;
        }
        return a * b;
    }

    private void parseExpression(String expression) {
        numbers = new ArrayList<>();
        operators = new ArrayList<>();
        int num = 0;
        for (int i = 0; i < expression.length(); i++) {
            char c = expression.charAt(i);
            if (Character.isDigit(c)) {
                num = num * 10 + (c - '0');
            } else {
                numbers.add(num);
                operators.add(c);
                num = 0;
            }
        }
        numbers.add(num);
    }
}

// time O(n^2 * 2^n)
// space O(n^2 * 2^n)